bookkeeper-server/src/main/java/org/apache/bookkeeper/client/CRC32CDigestManager.java

```diff
@@ -0,0 +1,65 @@
+package org.apache.bookkeeper.client;
+
+/*
+* Licensed to the Apache Software Foundation (ASF) under one
+* or more contributor license agreements.  See the NOTICE file
+* distributed with this work for additional information
+* regarding copyright ownership.  The ASF licenses this file
+* to you under the Apache License, Version 2.0 (the
+* "License"); you may not use this file except in compliance
+* with the License.  You may obtain a copy of the License at
+*
+*     http://www.apache.org/licenses/LICENSE-2.0
+*
+* Unless required by applicable law or agreed to in writing, software
+* distributed under the License is distributed on an "AS IS" BASIS,
+* WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+* See the License for the specific language governing permissions and
+* limitations under the License.
+*/
+
+import com.scurrilous.circe.crc.Sse42Crc32C;
+import io.netty.buffer.ByteBuf;
+import org.apache.commons.lang3.mutable.MutableBoolean;
+import org.apache.commons.lang3.mutable.MutableInt;
+import org.apache.pulsar.checksum.utils.Crc32cChecksum;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+class CRC32CDigestManager extends DigestManager {
+    static final Logger LOG = LoggerFactory.getLogger(CRC32CDigestManager.class);
+
+    private final ThreadLocal<MutableInt> currentCrc = ThreadLocal
+            .withInitial(() -> new MutableInt(0));
+    private final ThreadLocal<MutableBoolean> isNewCrc = ThreadLocal
+            .withInitial(() -> new MutableBoolean(true));
+
+    public CRC32CDigestManager(long ledgerId) {
+        super(ledgerId);
+        if (!Sse42Crc32C.isSupported()) {
+            LOG.error("Sse42Crc32C is not supported, will use less slower CRC32C implementation.");
+        }
+    }
+
+    @Override
+    int getMacCodeLength() {
+        return 4;
```

Comment:
@merlimat @dlg99 shall we stick to 8 like crc32 for consistency? assume we might use jdk9 crc32c in future.

https://docs.oracle.com/javase/9/docs/api/java/util/zip/CRC32C.html#getValue--

Reply:
@sijie Is there any guarantee that these will be interchangeable? i.e. jdk's one may  use actual full long and pulsar's one will have int + another int (i.e. 0) as a filler. So we won't be able to upgrade bookie if it has ledgers created with one crc32c implementation and bookie just switched to another. Safer way would be to have yet another digest type (CRC32C_J) and set it explicitly.

Reply:
@dlg99 I think it should be exchangeable. 

CRC32C is defined as exactly 4 bytes. But I think most of java versions returns a long, is to 1) have compatibility with java.util.zip.Checksum and 2) to avoid signedness issues with putting a 4-byte quantity in a signed int. I think we can stick to 4 bytes or 8 bytes, it doesn't really matter here.